src/sqlfluff/rules/L069.py

```diff
@@ -0,0 +1,175 @@
+        which cannot be rewritten into CAST.
+        This rule is disabled by default for Teradata because it supports different
+        type casting apart from CONVERT and ::
+        e.g DATE '2007-01-01', '9999-12-31' (DATE).
+
+    **Anti-pattern**
+
+    Using CONVERT or :: instead of CAST when ``allow_non_standard_type_casting``
+    config is set to ``False`` (default).
+
+    .. code-block:: sql
+
+        SELECT convert(varchar, 1) AS bar
+        FROM foo;
+
+        SELECT 1::varchar AS bar
+        FROM foo;
+
+    **Best practice**
+
+    Use ``CAST`` instead.
+    ``CAST`` is universally supported
+
+    .. code-block:: sql
+
+        SELECT CAST(1 AS varchar) AS bar,
+        FROM foo;
+
+    """
+
+    groups = ("all",)
+    config_keywords = ["allow_non_standard_type_casting"]
+    crawl_behaviour = SegmentSeekerCrawler({"function", "cast_expression"})
+
+    def _eval(self, context: RuleContext) -> Optional[LintResult]:
+        """Use ``CAST`` instead of ``CONVERT`` or ``::``."""
+        # Config type hints
+        self.allow_non_standard_type_casting: bool
+
+        # Rule disabled for teradata.
+        if context.dialect.name == "teradata":
+            return None
+
+        if not self.allow_non_standard_type_casting:
+            # Limit scope to CONVERT and casting_operator(::)
+            # CONVERT
+            if (
+                context.segment.is_type("function")
+                and context.segment.get_child("function_name").raw_upper == "CONVERT"
+            ):
+                # Get the content of CONVERT
+                convert_content = (
+                    FunctionalContext(context)
+                    .segment.children(sp.is_type("bracketed"))
+                    .children(
+                        sp.and_(
+                            sp.not_(sp.is_meta()),
+                            sp.not_(
+                                sp.is_type(
+                                    "start_bracket",
+                                    "end_bracket",
+                                    "whitespace",
+                                    "newline",
+                                    "comma",
+                                )
+                            ),
+                        )
+                    )
+                )
+
+                # we only care about 2-arguments convert
+                # some dialects allow an optional 3rd argument e.g TSQL
+                # which cannot be rewritten into CAST
+                if len(convert_content) > 2:
+                    return None
+
+                # get the expression
+                expression = convert_content[1].raw
+
+                # get the data type
+                datatype = convert_content[0].raw
+
+                # Create fix to replace ``CONVERT`` with ``CAST``.
+                fix = LintFix.replace(
+                    context.segment,
+                    [
+                        CodeSegment(
```

Comment:
Can you look into creating the new code using the individual segment types, rather than a single raw segment? During the fix process, we want to maintain a parse tree that matches what you'd get from parsing the fixed code. Otherwise, linting rules won't work well with one another.

Reply:
I am not sure how that would look. `cast_expression` comprises a single segment that is being changed. Also, something like `a::varchar` => `cast_expression` is being transformed into `cast(a as varchar)` => function, so I am not sure how we could use individual segment here.

Reply:
I mean, the rule heeds to create the same segments that would exist if you'd written the code by hand. You can use `sqlfluff parse` on a manually fixed file to figure out what it looks like. Other rules have examples of how to write the Python.

Reply:
Okay, could you point me to an example?

Reply:
L042 and L043 do this. Unfortunately, these techniques are not well documented, partly because they are not well established. Most other autofix rules fix individual segments or at least a "flat" structure. In this case, the fix involves multiple, nested segments.

Reply:
Also L032.

Reply:
Okay. Thank you, I am checking them.

Reply:
Fixed by [ebeb4ad](https://github.com/sqlfluff/sqlfluff/pull/3747/commits/ebeb4ada735810c3d050d690d93e0d8fcc64d4fa)